core/src/test/java/org/testcontainers/utility/AuthenticatedImagePullTest.java

```diff
@@ -0,0 +1,117 @@
+
+    @ClassRule
+    public static GenericContainer authenticatedRegistry = new GenericContainer(new ImageFromDockerfile()
+        .withDockerfileFromBuilder(builder -> {
+            builder.from("registry:2")
+                .run("htpasswd -Bbn testuser notasecret > /htpasswd")
+                .env("REGISTRY_AUTH", "htpasswd")
+                .env("REGISTRY_AUTH_HTPASSWD_PATH", "/htpasswd")
+                .env("REGISTRY_AUTH_HTPASSWD_REALM", "Test");
+        }))
+        .withExposedPorts(5000)
+        .waitingFor(new HttpWaitStrategy());
+
+    private static RegistryAuthLocator originalAuthLocatorSingleton;
+    private static DockerClient client;
+
+    private static String testRegistryAddress;
+    private static String testImageName;
+    private static String testImageNameWithTag;
+
+    @BeforeClass
+    public static void setUp() {
+        originalAuthLocatorSingleton = RegistryAuthLocator.instance();
+        client = DockerClientFactory.instance().client();
+
+        testRegistryAddress = authenticatedRegistry.getContainerIpAddress() + ":" + authenticatedRegistry.getFirstMappedPort();
+        testImageName = testRegistryAddress + "/alpine";
+        testImageNameWithTag = testImageName + ":latest";
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        RegistryAuthLocator.setInstance(originalAuthLocatorSingleton);
+        client.removeImageCmd(testImageNameWithTag).withForce(true).exec();
+    }
+
+    @Test
+    public void testThatAuthLocatorIsUsed() throws Exception {
+
+        final DockerImageName expectedName = new DockerImageName(testImageNameWithTag);
+        final AuthConfig authConfig = new AuthConfig()
+            .withUsername("testuser")
+            .withPassword("notasecret")
+            .withRegistryAddress("http://" + testRegistryAddress);
+
+        // Replace the RegistryAuthLocator singleton with our mock, for the duration of this test
+        final RegistryAuthLocator mockAuthLocator = Mockito.mock(RegistryAuthLocator.class);
+        RegistryAuthLocator.setInstance(mockAuthLocator);
+        when(mockAuthLocator.lookupAuthConfig(eq(expectedName), any()))
+            .thenReturn(authConfig);
+
+        // a push will use the auth locator for authentication, although that isn't the goal of this test
+        putImageInRegistry();
+
+        // actually start a container, which will require an authenticated pull
+        try (final GenericContainer container = new GenericContainer<>(testImageNameWithTag)
+            .withCommand("/bin/sh", "-c", "sleep 10")) {
+            container.start();
+
+            assertTrue("container started following an authenticated pull", container.isRunning());
+        }
+    }
+
+    private void putImageInRegistry() throws InterruptedException {
+        client.pullImageCmd("alpine:latest")
+            .exec(new PullImageResultCallback())
+            .awaitCompletion(1, TimeUnit.MINUTES);
+
+        final String id = client.listImagesCmd()
```

Comment:
there is `client.inspectImageCmd("alpine:latest").exec().getId()`

Reply:
Good point